File: mobile/androidApp/src/main/kotlin/dev/centraid/android/kit/MakeVaultSheet.kt
package dev.centraid.android.kit

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import dev.centraid.android.theme.centraidColor
import dev.centraid.android.theme.centraidType
import dev.centraid.design.CentraidGeometry

/**
 * THE MAKE-A-VAULT SHEET — WHAT `GatewaySheet` BECAME (#1029 §1).
 *
 * It was "connect this device to a gateway": a field to paste a pairing ticket
 * into, a "Pair this device" button, a "Sync now" button, and the last sentence
 * either of them produced. All three are gone with the plane behind them —
 * there is no ticket to redeem, no gateway to redeem it against, and no pass to
 * run. **The phone is the vault**, so what is left is the one act that
 * replaced them: make one.
 *
 * Same door as iOS `MakeVaultSheet`. Named apart from `HomeScreen`'s own private
 * `VaultSheet`, which is the SWITCHER: one lists the vaults this device holds,
 * this one adds to them. The sheet decides nothing — [onFound] is a
 * `HomeSession` call the activity owns.
 *
 * **R-SHELL-2:** [status] must name the foreground holding (or be empty), never
 * a vault this device has forgotten or switched away from.
 */
@Composable
public fun MakeVaultSheet(
    status: String,
    working: Boolean,
    onFound: (done: () -> Unit) -> Unit,
    onOpenTransferRules: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier
            .fillMaxWidth()
            .padding(CentraidGeometry.PAGE_MARGIN.dp)
            .padding(bottom = 24.dp)
            .testTag("vault-sheet"),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        Text(
            "Vault",
            style = centraidType("title"),
            color = centraidColor("text"),
        )
        Text(
            "Make a vault on this phone",
            style = centraidType("smallStrong"),
            color = centraidColor("text"),
        )
        TextButton(
            onClick = { if (!working) onFound {} },
            enabled = !working,
            modifier = Modifier
                .fillMaxWidth()
                .testTag("vault-found-button"),
        ) {
            Text(
                if (working) "Making…" else "Make a vault",
                color = centraidColor("link"),
            )
        }
        TextButton(
            onClick = onOpenTransferRules,
            modifier = Modifier
                .fillMaxWidth()
                .testTag("vault-transfer-rules"),
        ) {
            Text("Download settings", color = centraidColor("text"))
        }
        if (status.isNotEmpty()) {
            Text(
                status,
                style = centraidType("small"),
                color = centraidColor("textSoft"),
                modifier = Modifier.testTag("vault-status"),
            )
        }
    }
}

File: mobile/iosApp/Sources/MakeVaultSheet.swift
import SwiftUI

/// MAKE A VAULT ON THIS PHONE (#1029 §1).
///
/// The door behind Settings, and what `GatewaySheet` became. Named apart from
/// `HomeView`'s own private `VaultSheet`, which is the SWITCHER: one lists the
/// vaults this device holds, this one adds to them.
///
/// ## What it was, and why none of it survives
///
/// It was "connect this device to a gateway": a field to paste a pairing ticket
/// into, a "Pair this device" button, a "Sync now" button and the sentence
/// either produced. The ticket had no camera behind it on purpose — the
/// simulator has none and the scenarios this shell had to be testable against
/// all ran there — and `HomeSession.pair` judged nothing locally because only a
/// gateway could judge a ticket.
///
/// **The phone is the vault.** There is no ticket to paste, no gateway to
/// redeem it against and no pass to run, so all three controls are gone and one
/// takes their place: found a vault here.
///
/// ## Nothing here decides anything
///
/// Still true, and it is the reason this sheet is three controls and no state.
/// Whether a vault can be founded, and what it ends up called, are
/// `HomeSession.found` and, under it, the core. A shell that pre-judged either
/// would be a second opinion about a value only the vault can state.
struct MakeVaultSheet: View {
    @ObservedObject var shell: ShellModel
    @Environment(\.dismiss) private var dismiss
    @Environment(\.colorScheme) private var scheme

    @State private var working = false

    var body: some View {
        NavigationStack {
            Form {
                Section("Make a vault") {
                    Button(working ? "Making…" : "Make a vault on this phone") {
                        working = true
                        shell.found { working = false }
                    }
                    .disabled(working)
                    .accessibilityIdentifier("vault-found-button")

                    // THE LAST OUTCOME, IN A MEMBER'S WORDS. Never a hash,
                    // never a path, never a peer's error text — `gatewayStatus`
                    // carries only what the core already decided was showable.
                    if !shell.gatewayStatus.isEmpty {
                        Text(shell.gatewayStatus)
                            .centraidType("small")
                            .foregroundStyle(Theme.color("textSoft", scheme))
                            .accessibilityIdentifier("vault-status")
                    }
                }

                // THE TRANSFER RULES, ALWAYS REACHABLE (#1025 S4).
                //
                // The header's "N originals waiting for Wi-Fi" line is the
                // CONTEXTUAL door and cannot be the only one: on a device where
                // nothing is being withheld the line is empty, so a member who
                // wants to CHANGE the rule before it ever bites — set `manual`
                // on a new phone, say — would have no way in. A setting whose
                // only door appears once the setting has already cost
                // something is a setting a member cannot choose in advance.
                Section("Downloads") {
                    Button("Download settings") { shell.openTransferRules() }
                        .accessibilityIdentifier("vault-transfer-rules")
                }
            }
            .navigationTitle("Vault")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Done") { dismiss() }
                        .accessibilityIdentifier("vault-done")
                }
            }
        }
    }
}
